### ISSUE TITLE
Easy-effects accessibility with Orca

### ISSUE THREAD

**Issue by devinprater** [wwmm/easyeffects#1168]
Hello, I'm a blind Linux user, and am wondering if Easy-effects could be made more accessible with Orca? I have the version from Flathub, latest stable version, running on Fedora 34 with Orca version 40.0. Steps to reproduce:

* Turn on orca by running orca from the Alt + F2 run dialog.
* Open Personal > Assistive Technologies from your applications or system menu, or search for Assistive Technologies.
* Check the "Enable Assistive Technologies" checkbox in that dialog, and close it, saving or applying changes if needed.
* Open Easy-effects.
* You'll hear a lot of "GTK MM..." "list view" stuff.
* Arrowing down the list view shows applications.
* Tab to try to find effects lists. You'll hear more "GTKMM..." stuff, some switches, buttons, and tabs, but nothing about effects.
* Also, while in Easy-effects, speech is not stopped (the speech queue isn't flushed) when tabbing or doing anything else, but everywhere else it is.
If I could help in any other way, let me know. I'd love to use this, to help make sound a little better, or use BS2B to put the sound somewhere besides right in front of me, right at my ears, for a more comfortable listening experience over 8 hours per day.

**Comment by wwmm** [comment]
Things seem a little different here on Arch Linux. It does not seem I have a "Assistive Technologies" section. What you are using is the GNOME screen reader?

**Comment by devinprater** [comment]
Yes. If you're on Gnome, maybe look in the Control Center, or Control Panel.

**Comment by jbstep** [comment]
Orca is listed under Accessibility in KDE here in my Garuda system.

**Comment by vchernin** [comment]
From what I can tell what's missing here is better integration within EasyEffects itself? Orca itself is working since it's reading things out, the issue is Orca is being told to read information which is not relevant to the user.

I think [this GTK doc is relevant](https://docs.gtk.org/gtk4/section-accessibility.html). From what I understand better integration with Orca is kind of similar to what adding translations feels like? Is there a tool like Poedit but for accessibility/Orca?

**Comment by devinprater** [comment]
There is Accerciser: https://gitlab.gnome.org/GNOME/accerciser
Devin Prater
***@***.***
gemini://tilde.pink/~devinprater/

**Comment by Digitalone1** [comment]
If I have time, I can install accerciser tomorrow and see what is missing.

**Comment by Digitalone1** [comment]
Unfortunately I can't test this because I'm not able to run the speech dispatcher service.

The voice tells me there's an issue and I should check a log file and /var/log or local user directory, but there's no log related to speech-dispatcher.

When I try to start the service through systemd, it's immediately terminated. The status doesn't give me useful information:

```
Starting Speech-Dispatcher an high-level device independent layer for speech synthesis....
speechd: Speech Dispatcher 0.10.2 starting
speech-dispatcherd.service: Control process exited, code=exited, status=1/FAILURE
speech-dispatcherd.service: Failed with result 'exit-code'.
Failed to start Speech-Dispatcher an high-level device independent layer for speech synthesis..
``` 

Accerciser gives only information about which service is supported by the application. Not much help about how it has to be implemented.

**Comment by Digitalone1** [comment]
@devinprater do you have any hint to resolve this issue I have with speech dispatcher? I don't have any clue how to make it work.

**Comment by devinprater** [comment]
Make sure a speech engine, like espeak or rhvoice is installed. Did you
install orca? If so it should have installed a speech engine, but if not
try installing one.
Devin Prater
***@***.***
gemini://tilde.pink/~devinprater/

**Comment by Digitalone1** [comment]
@devinprater Ok, espeak was missing.

I can now listen and understand the issue. I hope to resolve it, but GTK documentation is not of much help since it does not give any example on how to implement a good accessible structure.

I will try to make some attempts playing with ui xml files.

**Comment by Digitalone1** [comment]
So I made a first test. Setting the accessible label property is reducing the "GTKMM" spelling, but not all widgets are working as expected.

What works:
* spinbuttons and scales are good: orca reproduces the label and the widget value
* simple buttons are good
* checkbuttons seems to work fine

What is not working
* for toggle buttons the label is reproduced, but there's no way to inform the user if the state is checked or unchecked - I tried also to modify the property on changes, but it's not working - I don't know if it's an issue related to orca or GTK4
* buttons that show a popover can't reproduce the label, they are presented and "GTK double button" - this should be an orca issue
* buttons with images are not well reproduced because the voice says "GTK image" and not always the label is reproduced - maybe a workaround could be making an image object as child and make it as "presentation role" so it should be ignored

I didn't test combobox, lists and other widgets completely, but I don't think they are well reproduced. For now I just made a first test implementation for application window ui and the maximizer which is one of the easiest plugin to deal with.

This needs lots of works and at last I think it won't fully work because GTK or orca miss something. At least if we provide the right information, it's better than nothing, hoping GTK will fix issue in the future.

@devinprater Could you test the master branch? Which plugins are you using now? We can help you to find a better preset if you need.

**Comment by Digitalone1** [comment]
@vchernin @wwmm has not merged it yet to master. The test is on my repository.

**Comment by vchernin** [comment]
> The test is on my repository.

The test build just takes it from your repo. It's not a problem to switch between them.

Oh, now that I reread your message I see you didn't intend for them to test your branch necessarily. My bad.

**Comment by devinprater** [comment]
Thanks so much for y'all's work. I'll test Monday when I'm at a Linux PC.
Devin Prater
***@***.***
gemini://tilde.pink/~devinprater/

**Comment by Digitalone1** [comment]
> Oh, now that I reread your message I see you didn't intend for them to test your branch directly necessarily. My bad.

For next monday @wwmm should merge it to master.

> FYI @joanmarie is the orca developer, maybe she can suggest something.

Sure, any help is appreciated.

**Comment by Digitalone1** [comment]
Made another merge request. At the moment only maximizer is working properly. Label property has been added to almost all spinboxes and scales, but plugins with comboboxes are not well handled because Orca can't reproduce them in a proper way. I'll try to improve other plugins and investigate on the other widgets in the next days.
